peachjam/adapters/adapters.py

```diff
@@ -0,0 +1,150 @@
+        """Update the document identified by some opaque id, returned by check_for_updates."""
+        raise NotImplementedError()
+
+    @classmethod
+    def name(cls):
+        return cls.__name__
+
+
+@plugins.register("ingestor-adapter")
+class IndigoAdapter(Adapter):
+    def check_for_updates(self, last_refreshed):
+        """Checks for documents updated since last_refreshed (which may be None), and returns a list
+        of document identifiers (expression FRBR URIs) which must be updated.
+        """
+        updated_docs_list = self.get_updated_documents(last_refreshed)
+        return [d["url"] for d in updated_docs_list]
+
+    def get_doc_list(self):
+        return self.client_get(self.url).json()["results"]
+
+    def get_updated_documents(self, last_refreshed):
+        if last_refreshed is None:
+            return self.get_doc_list()
+
+        return [
+            document
+            for document in self.get_doc_list()
+            if parser.parse(document["updated_at"]) > last_refreshed
+        ]
+
+    def update_document(self, url):
+        from countries_plus.models import Country
+        from languages_plus.models import Language
+
+        from africanlii.models import Legislation
+        from peachjam.models import Locality, Work
+
+        logger.info(f"Updating document ... {url}")
+
+        document = self.client_get(f"{url}.json").json()
+        frbr_uri = document["frbr_uri"]
+        title = document["title"]
+        expression_frbr_uri = document["expression_frbr_uri"]
+        toc_json = self.get_toc_json(url)
+        content_html = self.client_get(url + ".html").text
+        jurisdiction = Country.objects.get(iso__iexact=document["country"])
+        language = Language.objects.get(iso_639_3__iexact=document["language"])
+        locality = Locality.objects.get(code=document["locality"])
+        work, _ = Work.objects.update_or_create(
+            frbr_uri=frbr_uri, defaults={"title": title}
+        )
+
+        field_data = {
+            "title": title,
+            "created_at": document["created_at"],
+            "updated_at": document["updated_at"],
+            "work_frbr_uri": frbr_uri,
+            "date": document["expression_date"],
+            "content_html_is_akn": True,
+            "source_url": document["publication_document"]["url"]
+            if document["publication_document"]
+            else None,
+            "metadata_json": document,
+            "jurisdiction": jurisdiction,
+            "locality": locality,
+            "language": language,
+            "toc_json": toc_json,
+            "content_html": content_html,
+            "work": work,
+        }
+
+        doc, _ = Legislation.objects.update_or_create(
+            expression_frbr_uri=expression_frbr_uri, defaults={**field_data}
+        )
+        if document["publication_document"]:
+            self.download_source_file(document["publication_document"], doc)

```

Comment:
I'd like to treat this a bit differently. The idea behind peachjam's sourcefile is that it's the "original" file which the user can download.

With Indigo, the publication document is strictly the original source file, but often that is a huge gazette that doesn't have all amendments.

Instead, we need to download the rendered PDF from Indigo and attach that as the source file.

To get the PDF, fetch `document["url"] + ".pdf"` much like you do for the publication document below, and then stash that file.

Reply:
Changed. I am assuming that the api always generates a pdf at that url and so it will always be present?

Reply:
Yes, correct.